unstructured_ingest/v2/processes/connectors/ibm_watsonx_data.py

```diff
@@ -0,0 +1,298 @@
+    namespace: str = Field(description="Namespace name")
+    table: str = Field(description="Table name")
+    max_retries: int = Field(
+        default=5, description="Maximum number of retries to upload data", ge=2, le=10
+    )
+    record_id_key: str = Field(
+        default=RECORD_ID_LABEL,
+        description="Searchable key to find entries for the same record on previous runs",
+    )
+
+    @property
+    def table_identifier(self) -> Tuple[str, str]:
+        return (self.namespace, self.table)
+
+
+@dataclass
+class IbmWatsonxDataUploader(Uploader):
+    connection_config: IbmWatsonxDataConnectionConfig
+    upload_config: IbmWatsonxDataUploaderConfig
+    connector_type: str = CONNECTOR_TYPE
+
+    def precheck(self) -> None:
+        with self.connection_config.get_catalog() as catalog:
+            if not catalog.namespace_exists(self.upload_config.namespace):
+                raise DestinationConnectionError(
+                    f"Namespace '{self.upload_config.namespace}' does not exist"
+                )
+            if not catalog.table_exists(self.upload_config.table_identifier):
+                raise DestinationConnectionError(
+                    f"Table '{self.upload_config.table}' does not exist in namespace '{self.upload_config.namespace}'"  # noqa: E501
+                )
+
+    @contextmanager
+    def get_table(self) -> Generator["Table", None, None]:
+        with self.connection_config.get_catalog() as catalog:
+            table = catalog.load_table(self.upload_config.table_identifier)
+            yield table
+
+    @requires_dependencies(["pyarrow"], extras="ibm-watsonx-data")
+    def _get_data_table(self, table: "Table", path: Path) -> Any:
+        import pyarrow as pa
+
+        df = get_data_df(path)
+        table_column_names = table.schema().column_names
+        df_column_names = list(df.columns.values)
+        # Only upload columns that are common between the table and the dataframe
+        common_columns = list(set(table_column_names).intersection(df_column_names))
+        if not common_columns:
+            raise ValueError(
+                "Iceberg table schema doesn't contain proper columns"
+                f"Expected columns: {df_column_names}"
+                f"Found columns: {table_column_names}"
+            )
+        return pa.Table.from_pandas(df[common_columns])
+
+    @requires_dependencies(["pyiceberg"], extras="ibm-watsonx-data")
+    def upload_data(self, table: "Table", data_table: Any, file_data: FileData) -> None:
+        """
+        Uploads data to a specified table.
+
+        This method handles the process of uploading data to a table, including
+        deleting existing records if necessary and appending new data. It uses
+        a transaction wrapper to ensure the operation is retried if it fails.
+        """
+        from pyiceberg.expressions import EqualTo
+
+        def _upload_data(transaction: "Transaction", data_table: Any, file_data: FileData) -> None:
```

Comment:
Should this be typed as a dataframe? 
```suggestion
        def _upload_data(transaction: "Transaction", data_table: pd.DataFrame, file_data: FileData) -> None:
```

Reply:
changed